File: observer-pattern/c/src/concrete_observer.c
#include "func.h"

void concrete_observer_set_subject(ConcreteObserver *observer, Subject *subject)
{
  observer->subject = subject;
}

//  观察者更新事件，主题类有新发布时会批量调用，而无需逐个通知
void concrete_observer_update(ConcreteObserver *observer, char *content)
{
  printf("\r\n ConcreteObserver::update() [subject->name = %s content = %s]", observer->subject->name, content);
}

// 给观察者绑定主题，同时把观察者添加到主题列表
ConcreteObserver *concrete_observer_constructor(char *name)
{
  printf("\r\n concrete_observer_constructor() [构建ConcreteObserver]");
  Observer *observer = (Observer *)malloc(sizeof(Observer));
  ConcreteObserver *concrete_observer = (ConcreteObserver *)observer;
  strncpy(observer->name, name, 50);
  concrete_observer->set_subject = &concrete_observer_set_subject;
  concrete_observer->update = &concrete_observer_update;

  return concrete_observer;
}

File: observer-pattern/java/src/ConcreteObserver.java
package src;

// 具体的观察者实现类，也可以看成订阅者，关联对应的主题类。
// 不同的观察者也可以对应多个主题
public class ConcreteObserver implements ObserverAPI {

   public Subject subject;

   // 给观察者绑定主题，同时把观察者添加到主题列表
   public ConcreteObserver(Subject subject) {
      this.subject = subject;
      this.subject.register((ObserverAPI) this);
   }

   // 观察者更新事件，主题类有新发布时会批量调用，而无需逐个通知
   public void update(String content) {
      System.out.println(String.format("%s::update() [subject.name = %s content = %s]",
            this.getClass().getName(),
            this.subject.getName(), content));
   }
}